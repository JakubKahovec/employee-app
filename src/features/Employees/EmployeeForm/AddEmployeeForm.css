.AddEmployeeForm {
  display: grid;
  grid-row: auto;
  padding: 20px;
  grid-auto-columns: auto;
}

