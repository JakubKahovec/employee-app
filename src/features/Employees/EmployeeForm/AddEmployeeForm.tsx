import { Employee } from "AppModel";
import { Alert, Button, Form, Input, InputNumber, Select } from "antd";
import { routerActions } from "connected-react-router";
import * as React from "react";
import { connect } from "react-redux";
import { RootState } from "typesafe-actions";
import { routes } from "../../../Routes";
import * as actions from "../EmployeesActions";
import "./AddEmployeeForm.css";

export class AddEmployeeForm extends React.Component<Props> {
  render() {
    return (
      <div className="AddEmployeeForm">
        <h3>Add new employee</h3>
        <Form {...layout} name="basic" initialValues={{}} onFinish={this.onFinish}>
          <Form.Item label="Name" name="name" rules={[{ required: true, message: "Please input your name" }]}>
            <Input placeholder="Employee full name" />
          </Form.Item>
          <Form.Item
            label="Job Title"
            name="jobTitle"
            rules={[{ required: true, message: "Please input your job title" }]}
          >
            <Input placeholder="Job title" />
          </Form.Item>
          <Form.Item name="gender" label="Gender" rules={[{ required: true, message: "Please select gender" }]}>
            <Select placeholder="Select a gender" allowClear>
              <Option value="Male">male</Option>
              <Option value="Female">female</Option>
            </Select>
          </Form.Item>
          <Form.Item
            name="tenure"
            label="Tenure"
            rules={[{ required: true, type: "number", min: 1, max: 99, message: "Please select tenure" }]}
          >
            <InputNumber placeholder="Tenure" />
          </Form.Item>
          <Form.Item {...tailLayout}>
            <Button onClick={this.onCancel}>Cancel</Button>
            <Button loading={this.props.isSavingEmployee} data-testid="submit-button" type="primary" htmlType="submit">
              Submit
            </Button>
          </Form.Item>
        </Form>
        {this.props.saveEmployeeError ? <Alert type="error" closable message={this.props.saveEmployeeError} /> : null}
      </div>
    );
  }

  onFinish = (values: any) => {
    this.props.saveEmployee(values as Employee);
  };

  onCancel = () => {
    this.props.goTo(routes.dashboard);
  };
}

const { Option } = Select;

const layout = {
  labelCol: { span: 8 },
  wrapperCol: { span: 8 },
};
const tailLayout = {
  wrapperCol: { offset: 8, span: 8 },
};

const mapStateToProps = (state: RootState) => ({
  isSavingEmployee: state.employees.isSavingEmployees,
  saveEmployeeError: state.employees.saveEmployeeError,
});
const dispatchProps = {
  saveEmployee: actions.saveEmployeeAsync.request,
  goTo: routerActions.push,
};

type Props = ReturnType<typeof mapStateToProps> & typeof dispatchProps;

export default connect(mapStateToProps, dispatchProps)(AddEmployeeForm);
