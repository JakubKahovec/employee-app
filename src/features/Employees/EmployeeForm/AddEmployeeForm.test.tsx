import { render } from "@testing-library/react";
import * as React from "react";
import { act } from "react-dom/test-utils";
import { createRenderer } from "react-test-renderer/shallow";
import { AddEmployeeForm } from "./AddEmployeeForm";

describe("AddEmployeeForm", () => {

  it("renders correctly", () => {
    const renderer = createRenderer();
    const tree = renderer.render(
      <AddEmployeeForm goTo={jest.fn()} saveEmployee={jest.fn()} saveEmployeeError="" isSavingEmployee={false} />,
    );
    expect(tree).toMatchSnapshot();
  });

  it("renders a save error message", () => {
    const { getByText } = render(
      <AddEmployeeForm
        goTo={jest.fn()}
        saveEmployee={jest.fn()}
        saveEmployeeError="Failed to save"
        isSavingEmployee={false}
      />,
    );
    const errorMsg = getByText(/Failed to save/i);
    expect(errorMsg).toBeInTheDocument();
  });

  it("prevents to submit on errors", async () => {
    const saveEmployee = jest.fn();
    const { container, getByTestId } = render(
      <AddEmployeeForm goTo={jest.fn()} saveEmployee={saveEmployee} saveEmployeeError="" isSavingEmployee={false} />,
    );
    const submitBtn = getByTestId("submit-button");
    jest.useFakeTimers();
    await act(async () => {
      submitBtn.dispatchEvent(new MouseEvent("click", { bubbles: true }));
      jest.runAllTimers();
    });
    const errorSelectName = container.querySelector(".ant-form-item-has-error");
    expect(saveEmployee).toBeCalledTimes(0);
    expect(errorSelectName).toBeInTheDocument();
  });
});
