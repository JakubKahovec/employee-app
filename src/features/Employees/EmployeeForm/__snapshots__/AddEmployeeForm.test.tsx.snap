// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddEmployeeForm renders correctly 1`] = `
<div
  className="AddEmployeeForm"
>
  <h3>
    Add new employee
  </h3>
  <ForwardRef(InternalForm)
    initialValues={Object {}}
    labelCol={
      Object {
        "span": 8,
      }
    }
    name="basic"
    onFinish={[Function]}
    wrapperCol={
      Object {
        "span": 8,
      }
    }
  >
    <FormItem
      label="Name"
      name="name"
      rules={
        Array [
          Object {
            "message": "Please input your name",
            "required": true,
          },
        ]
      }
    >
      <Input
        placeholder="Employee full name"
        type="text"
      />
    </FormItem>
    <FormItem
      label="Job Title"
      name="jobTitle"
      rules={
        Array [
          Object {
            "message": "Please input your job title",
            "required": true,
          },
        ]
      }
    >
      <Input
        placeholder="Job title"
        type="text"
      />
    </FormItem>
    <FormItem
      label="Gender"
      name="gender"
      rules={
        Array [
          Object {
            "message": "Please select gender",
            "required": true,
          },
        ]
      }
    >
      <Select
        allowClear={true}
        bordered={true}
        choiceTransitionName="zoom"
        placeholder="Select a gender"
        transitionName="slide-up"
      >
        <Option
          value="Male"
        >
          male
        </Option>
        <Option
          value="Female"
        >
          female
        </Option>
      </Select>
    </FormItem>
    <FormItem
      label="Tenure"
      name="tenure"
      rules={
        Array [
          Object {
            "max": 99,
            "message": "Please select tenure",
            "min": 1,
            "required": true,
            "type": "number",
          },
        ]
      }
    >
      <InputNumber
        placeholder="Tenure"
        step={1}
      />
    </FormItem>
    <FormItem
      wrapperCol={
        Object {
          "offset": 8,
          "span": 8,
        }
      }
    >
      <Button
        block={false}
        ghost={false}
        htmlType="button"
        loading={false}
        onClick={[Function]}
      >
        Cancel
      </Button>
      <Button
        block={false}
        data-testid="submit-button"
        ghost={false}
        htmlType="submit"
        loading={false}
        type="primary"
      >
        Submit
      </Button>
    </FormItem>
  </ForwardRef(InternalForm)>
</div>
`;
